common/kvstore/src/main/java/org/apache/spark/util/kvstore/LevelDBIterator.java

```diff
@@ -280,4 +285,23 @@ static int compare(byte[] a, byte[] b) {
     return a.length - b.length;
   }
 
+  private record ResourceCleaner(DBIterator dbIterator, LevelDB levelDB) implements Runnable {
+    @Override
+    public void run() {
+      levelDB.getIteratorTracker().removeIf(ref -> {
+        LevelDBIterator<?> levelDBIterator = ref.get();
+        return levelDBIterator != null && dbIterator.equals(levelDBIterator.it);
```

Comment:
When is `levelDBIterator` going to be `null` ?
Also, instead of `equals` here shouldnt it not be instance check (`==`) ?

+CC @LuciferYang, looks like there is a leak in the DB, iterator's weak ref is never cleaned up from the tracker ?

Reply:
```
looks like there is a leak in the DB, iterator's weak ref is never cleaned up from the tracker ?
```

I think this should be a old problem, and we have also performed null checks in the close method of RocksDB.

https://github.com/apache/spark/blob/7d7afb06f682c10f3900eb8adeab9fad6d49cb24/common/kvstore/src/main/java/org/apache/spark/util/kvstore/RocksDB.java#L341-L348

https://github.com/apache/spark/blob/7d7afb06f682c10f3900eb8adeab9fad6d49cb24/common/kvstore/src/main/java/org/apache/spark/util/kvstore/LevelDB.java#L307-L315

If an iterator is no longer referenced and is not manually closed, then if GC occurs at this time, the iteratorTracker will still hold the Ref, but Ref#get is null.

Should we change this condition to 

```
return levelDBIterator == null || dbIterator.equals(levelDBIterator.it); 
```

Can this also clean up the invalid weakref in passing?

Reply:
With respect to `WeakReference` accumulating in that list, agree - that is an existing problem - this PR is not changing it.
I noticed it while reviewing, and so pinged you @LuciferYang :-)

Reply:
w.r.t change in logic ... in current code, I dont think `levelDB.getIteratorTracker()` will ever contain a `null` within it (the ref it points to can be `null`).

Given this is enhancement for `finalize` -> `Cleaner`, it is fine to simply use existing semantics and change this condition to simply `dbIterator == levelDBIterator.it`, thoughts ?

Reply:
Agree +1, we can fix it in follow up

Reply:
> Given this is enhancement for finalize -> Cleaner, it is fine to simply use existing semantics and change this condition to simply dbIterator == levelDBIterator.it, thoughts ?

hmm... `levelDBIterator` is `ref.get`, if `levelDBIterator` is null, `levelDBIterator.it` will throw NPE ...

Reply:
I simplified the condition of rocksdb part  into `dbIterator == rocksDBIterator.it` for local testing, there will be many test failures.

Reply:
You are right, I missed that levelDBIterator is itself ref.get ... sigh, sorry about this !

Reply:
sorry for late, do we need to change anything?